File: lib/src/main/java/com/otaliastudios/transcoder/internal/TrackMap.java
package com.otaliastudios.transcoder.internal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.otaliastudios.transcoder.common.TrackType;

import java.util.HashMap;
import java.util.Map;

/**
 * An utility class for storing data relative to a single {@link TrackType}
 * in a map, with handy nullability annotations.
 *
 * @param <T> the map type
 */
public class TrackMap<T> {

    public TrackMap() {
    }

    public TrackMap(@NonNull T videoValue, @NonNull T audioValue) {
        set(TrackType.AUDIO, audioValue);
        set(TrackType.VIDEO, videoValue);
    }

    private final Map<TrackType, T> map = new HashMap<>();

    public void set(@NonNull TrackType type, @Nullable T value) {
        map.put(type, value);
    }

    public void setAudio(@Nullable T value) {
        set(TrackType.AUDIO, value);
    }

    public void setVideo(@Nullable T value) {
        set(TrackType.VIDEO, value);
    }

    @Nullable
    public T get(@NonNull TrackType type) {
        return map.get(type);
    }

    @Nullable
    public T getAudio() {
        return get(TrackType.AUDIO);
    }

    @Nullable
    public T getVideo() {
        return get(TrackType.VIDEO);
    }

    @NonNull
    public T require(@NonNull TrackType type) {
        //noinspection ConstantConditions
        return map.get(type);
    }

    @NonNull
    public T requireAudio() {
        return require(TrackType.AUDIO);
    }

    @NonNull
    public T requireVideo() {
        return require(TrackType.VIDEO);
    }

    public boolean has(@NonNull TrackType type) {
        return map.containsKey(type);
    }

    public boolean hasAudio() {
        return has(TrackType.AUDIO);
    }

    public boolean hasVideo() {
        return has(TrackType.VIDEO);
    }
}

File: lib/src/main/java/com/otaliastudios/transcoder/internal/utils/TrackMap.kt
package com.otaliastudios.transcoder.internal.utils

import com.otaliastudios.transcoder.common.TrackType

interface TrackMap<T> : Iterable<T> {

    operator fun get(type: TrackType): T
    val video get() = get(TrackType.VIDEO)
    val audio get() = get(TrackType.AUDIO)

    fun has(type: TrackType): Boolean
    val hasVideo get() = has(TrackType.VIDEO)
    val hasAudio get() = has(TrackType.AUDIO)

    fun getOrNull(type: TrackType) = if (has(type)) get(type) else null
    fun videoOrNull() = getOrNull(TrackType.VIDEO)
    fun audioOrNull() = getOrNull(TrackType.AUDIO)

    val size get() = listOfNotNull(videoOrNull(), audioOrNull()).size

    override fun iterator() = listOfNotNull(videoOrNull(), audioOrNull()).iterator()
}

interface MutableTrackMap<T> : TrackMap<T> {
    operator fun set(type: TrackType, value: T?)

    fun reset(video: T?, audio: T?) {
        set(TrackType.VIDEO, video)
        set(TrackType.AUDIO, audio)
    }

    override var audio: T
        get() = super.audio
        set(value) = set(TrackType.AUDIO, value)

    override var video: T
        get() = super.video
        set(value) = set(TrackType.VIDEO, value)
}

fun <T> trackMapOf(default: T?) = trackMapOf(default, default)

fun <T> trackMapOf(video: T?, audio: T?): TrackMap<T> = DefaultTrackMap(video, audio)

fun <T> mutableTrackMapOf(default: T?) = mutableTrackMapOf(default, default)

fun <T> mutableTrackMapOf(video: T? = null, audio: T? = null): MutableTrackMap<T> = DefaultTrackMap(video, audio)

private class DefaultTrackMap<T>(video: T?, audio: T?) : MutableTrackMap<T> {
    private val map = mutableMapOf(TrackType.VIDEO to video, TrackType.AUDIO to audio)
    override fun get(type: TrackType): T = requireNotNull(map[type])
    override fun has(type: TrackType) = map[type] != null
    override fun set(type: TrackType, value: T?) {
        map[type] = value
    }
}

